### PR TITLE
encryption: abstract encryption-config deployment mechanism

### PR DESCRIPTION
**Pull request by sttts**
Prerequisite for
- integration tests with fake deployement
- openshift-apiserver support

### CODE REVIEW

File: pkg/operator/encryption/controllers.go [hunk 1]
```diff
@@ -9,87 +9,76 @@ import (
...
 	// TODO: update the eventHandlers used by the controllers to ignore components that do not match their own
-	encryptionSecretSelector := metav1.ListOptions{LabelSelector: secrets.EncryptionKeySecretsLabel + "=" + targetNamespace}
-
-	if err := resourceSyncer.SyncSecret(
-		resourcesynccontroller.ResourceLocation{Namespace: targetNamespace, Name: encryptionconfig.EncryptionConfSecretName},
-		resourcesynccontroller.ResourceLocation{Namespace: operatorclient.GlobalMachineSpecifiedConfigNamespace, Name: targetNamespace},
-	); err != nil {
-		return nil, err
-	}
+	encryptionSecretSelector := metav1.ListOptions{LabelSelector: secrets.EncryptionKeySecretsLabel + "=" + component}
```
Comment (mfojtik) [inline pkg/operator/encryption/controllers.go]: I like https://github.com/openshift/library-go/blob/master/pkg/operator/staticpod/controller/installerstate/installer_state_controller.go#L82 more

File: pkg/operator/encryption/deployer/staticpod.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+// as encryption-config. From there it is revisioned and deployed to the static pods. The last
+// deployed encryption config is read from encryption-config-<revision>.
+func NewStaticPodDeployer(
+	targetNamespace string,
+	informers operatorv1helpers.KubeInformersForNamespaces,
+	resourceSyncer resourcesynccontroller.ResourceSyncer,
+	podClient corev1client.PodsGetter,
+	secretClient corev1client.SecretsGetter,
+) (*StaticPodDeployer, error) {
+	if resourceSyncer != nil {
```
Comment (mfojtik) [inline pkg/operator/encryption/deployer/staticpod.go]: what is the case when resourceSyncer is nil?

File: pkg/operator/encryption/deployer/staticpod.go [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+	return &StaticPodDeployer{
+		podClient:                podClient.Pods(targetNamespace),
+		secretClient:             secretClient.Secrets(targetNamespace),
+		targetNamespaceInformers: informers.InformersFor(targetNamespace),
+	}, nil
+}
+
+// DeployedEncrytionConfigSecret returns the deployed encryption config and whether all
+// instances of the operand have acknowledged it.
+func (d *StaticPodDeployer) DeployedEncrytionConfigSecret() (secret *corev1.Secret, converged bool, err error) {
```
Comment (mfojtik) [inline pkg/operator/encryption/deployer/staticpod.go]: typo: DeployedEncryptionConfigSecret

File: pkg/operator/encryption/deployer/staticpod.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	targetSecretsInformer.AddEventHandler(handler)
+
+	return []cache.InformerSynced{
+		targetPodInformer.HasSynced,
+		targetSecretsInformer.HasSynced,
+	}
+}
+
+// revisionLabel is used to find the current revision for a given API server.
+const revisionLabel = "revision"
```
Comment (mfojtik) [inline pkg/operator/encryption/deployer/staticpod.go]: Mo will be proud :-)

File: pkg/operator/encryption/deployer/staticpod.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+// Converged can be defined as:
+//   1. All running pods are ready and at the same revision
+//   2. There are no pending or unknown pods
+//   3. All succeeded and failed pods have revisions that are before the running pods
+// Once a converged revision has been determined, it can be used to determine
+// what encryption config state has been successfully observed by the API servers.
+// It assumes that podClient is doing live lookups against the cluster state.
+func getAPIServerRevisionOfAllInstances(podClient corev1client.PodInterface) (string, error) {
+	// do a live list so we never get confused about what revision we are on
+	apiServerPods, err := podClient.List(metav1.ListOptions{LabelSelector: "apiserver=true"})
```
Comment (mfojtik) [inline pkg/operator/encryption/deployer/staticpod.go]: same comment as above

File: pkg/operator/encryption/deployer/staticpod.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+		}
+		if failedRevisionNum > revisionNum { // TODO can this dead lock?
+			return "", fmt.Errorf("api server has failed revision %v which is newer than running revision %v", failedRevisionNum, revisionNum)
+		}
+	}
+
+	return revision, nil
+}
+
+func podReady(pod corev1.Pod) bool {
```
Comment (mfojtik) [inline pkg/operator/encryption/deployer/staticpod.go]: feels like very common helper

File: pkg/operator/encryption/encryptionconfig/config.go [hunk 7]
```diff
@@ -72,11 +55,11 @@ func getGRsActualKeys(encryptionConfig *apiserverconfigv1.EncryptionConfiguratio
 	for _, resourceConfig := range encryptionConfig.Resources {
 		// resources should be a single group resource
 		if len(resourceConfig.Resources) != 1 {
-			klog.Infof("skipping invalid encryption config for resource %s", resourceConfig.Resources)
+			klog.Warningf("skipping invalid encryption config for resource %s", resourceConfig.Resources)
```
Comment (mfojtik) [inline pkg/operator/encryption/encryptionconfig/config.go]: note to myself: event worthy

File: pkg/operator/starter.go [hunk 8]
```diff
@@ -148,15 +149,16 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
 		return err
 	}
 
+	deployer, err := encryptiondeployer.NewStaticPodDeployer(operatorclient.TargetNamespace, kubeInformersForNamespaces, resourceSyncController, kubeClient.CoreV1(), kubeClient.CoreV1())
```
Comment (mfojtik) [inline pkg/operator/starter.go]: check the err ?

File: pkg/operator/encryption/deployer/staticpod.go (hunk 2 above)
Comment (sttts) [reply pkg/operator/encryption/deployer/staticpod.go]: in tests we don't want that

File: pkg/operator/encryption/deployer/staticpod.go (hunk 3 above)
Comment (sttts) [reply pkg/operator/encryption/deployer/staticpod.go]: my fingers don't like the word encryption.

Comment (sttts) [reply pkg/operator/encryption/deployer/staticpod.go]: fixed

File: pkg/operator/encryption/controllers.go (hunk 1 above)
Comment (sttts) [reply pkg/operator/encryption/controllers.go]: this is old code. Feel free to open another PR

File: pkg/operator/encryption/deployer/staticpod.go (hunk 4 above)
Comment (sttts) [reply pkg/operator/encryption/deployer/staticpod.go]: this is his code, just moved.

File: pkg/operator/encryption/deployer/staticpod.go (hunk 5 above)
Comment (sttts) [reply pkg/operator/encryption/deployer/staticpod.go]: just moved code

File: pkg/operator/encryption/deployer/staticpod.go (hunk 6 above)
Comment (sttts) [reply pkg/operator/encryption/deployer/staticpod.go]: just moved code

File: pkg/operator/starter.go (hunk 8 above)
Comment (sttts) [reply pkg/operator/starter.go]: done

File: pkg/operator/encryption/deployer/staticpod.go (hunk 2 above)
Comment (sttts) [reply pkg/operator/encryption/deployer/staticpod.go]: commented

**Comment by sttts** [comment]
Registry dead?

/retest

**Pull request merged**